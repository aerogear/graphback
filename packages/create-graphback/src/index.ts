#!/usr/bin/env node

import * as yargs from 'yargs';

import { init } from './init/components';

type Params = {
  name?: string;
  templateName?: string;
  templateUrl?: string;
  filter?: string;
};

const command = '$0 <name>';

export const desc = 'Create Graphback project from available templates';

// tslint:disable-next-line: typedef
export const builder = (args: yargs.Argv) => {
  args.positional('name', {
    describe: 'Project name',
    type: 'string'
  });
  args.option('templateName', {
    describe: 'Name of the predefined template',
    type: 'string'
  });
  args.option('templateUrl', {
    describe:
      'GitHub URL of the template. For example (https://github.com/wtrocki/graphql-serve-example#master)',
    type: 'string'
  });
  args.option('filter', {
    describe:
      'A filter option used to limit the number of displayed templates that only matches it. For example "postgres" will only display templates that have a Postgres backed',
    type: 'string'
  });
};

export async function handler({
  name,
  templateName,
  templateUrl,
  filter
}: Params) {
  await init(name, templateName, templateUrl, filter);
}

if (require.main === module) {
  // eslint-disable-next-line no-unused-expressions
  yargs
    .command(command, desc, builder, handler)
    .example(
      '$0 my-awesome-project',
      'creates the "my-awesome-project" in the current working directory'
    )
    .demandCommand(1)
    .strict()
    .recommendCommands()
    .help()
    .alias('h', 'help')
    .version()
    .alias('v', 'version').argv;
}
