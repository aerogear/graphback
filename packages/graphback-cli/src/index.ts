#!/usr/bin/env node

import * as yargs from 'yargs';

if (require.main === module) {
  // eslint-disable-next-line no-unused-expressions
  yargs
    .commandDir('commands')
    .demandCommand(1)
    .strict()
    .recommendCommands()
    .help()
    .alias('h', 'help')
    .version()
    .alias('v', 'version')
    .argv;
}
