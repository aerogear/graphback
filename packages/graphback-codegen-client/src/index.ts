export * from './ClientDocuments'
export * from './ClientCRUDPlugin'