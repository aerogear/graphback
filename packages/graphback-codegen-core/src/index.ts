 
// Top level API
export * from './api'
export * from './graphql'
