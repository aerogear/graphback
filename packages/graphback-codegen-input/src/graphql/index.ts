
export * from "./GraphQLContextCreator"
export * from "./graphqlUtils"