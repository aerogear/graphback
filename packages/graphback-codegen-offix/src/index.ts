export * from './OffixPlugin'
export * from './OffixRuntimeData'


//Required for plugins
export { OffixPlugin as Plugin } from './OffixPlugin'