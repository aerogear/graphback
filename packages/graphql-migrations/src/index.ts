export { migrateDB, MigrateOptions } from './migrate'
export { generateAbstractDatabase } from './abstract/generateAbstractDatabase'
export { computeDiff } from './diff/computeDiff'
export { read } from './connector/read'
export { write } from './connector/write'
export { MigratePlugin, WriteParams } from './plugin/MigratePlugin'
export * from './plugin/MigrateOperationFilter'

