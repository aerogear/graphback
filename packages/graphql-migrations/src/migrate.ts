import { buildSchema, GraphQLSchema } from 'graphql'
import * as Knex from 'knex';
import { generateAbstractDatabase, ScalarMap } from './abstract/generateAbstractDatabase'
import { read } from './connector/read'
import { write } from './connector/write'
import { computeDiff } from './diff/computeDiff'
import { MigrationResults } from './diff/Operation'
import { MigrateOperationFilter } from './plugin/MigrateOperationFilter';
import { MigratePlugin } from './plugin/MigratePlugin'

export interface MigrateOptions {
  /**
   * Table schema: `<schemaName>.<tableName>`.
   */
  dbSchemaName?: string
  /**
   * Table name prefix: `<prefix><tableName>`.
   */
  dbTablePrefix?: string
  /**
   * Column name prefix: `<prefix><columnName>`.
   */
  dbColumnPrefix?: string
  /**
   * Overwrite table and column comments (not supported in some databases).
   */
  updateComments?: boolean
  /**
   * Custom Scalar mapping
   */
  scalarMap?: ScalarMap | undefined
  /**
   * Map scalar/enum lists to json column type by default.
   */
  mapListToJson?: boolean
  /**
   * List of graphql-migrations plugins
   */
  plugins?: MigratePlugin[],
  /**
   * Display debug information
   */
  debug?: boolean
  /**
   * Remove directives from the GraphQLSchema
   */
  removeDirectivesFromSchema?: boolean

  /**
   * Method that can be used to filter out operations that we do not want to execute.
   * For example if we want to prevent deletion of the tables filter can remove `table.drop` operations
   * from array
   */
  operationFilter?: MigrateOperationFilter
}

export const defaultOptions: MigrateOptions = {
  dbSchemaName: 'public',
  dbTablePrefix: '',
  dbColumnPrefix: '',
  updateComments: false,
  scalarMap: undefined,
  mapListToJson: true,
  plugins: [],
  debug: false,
  removeDirectivesFromSchema: true,
}

export async function migrateDB(
  config: Knex.Config,
  schema: GraphQLSchema | string,
  options: MigrateOptions = {},
): Promise<MigrationResults> {
  //Default options
  const finalOptions = {
    ...defaultOptions,
    ...options
  }

  if (config.client === 'sqlite3') {
    finalOptions.dbSchemaName = '';
  }

  if (finalOptions.debug) {
    config = {
      ...config,
      debug: true,
    }
  }

  //Read current
  const existingAdb = await read(
    config,
    finalOptions.dbSchemaName,
    finalOptions.dbTablePrefix,
    finalOptions.dbColumnPrefix,
  )

  let finalSchema: GraphQLSchema;
  if (typeof schema === "string") {
    finalSchema = buildSchema(schema);
  } else {
    finalSchema = schema
  }

  //Generate new
  const newAdb = await generateAbstractDatabase(finalSchema, {
    scalarMap: finalOptions.scalarMap,
    mapListToJson: finalOptions.mapListToJson
  })

  if (finalOptions.debug) {
    // eslint-disable-next-line no-console
    console.log('BEFORE', JSON.stringify(existingAdb.tables, undefined, 2))
    // eslint-disable-next-line no-console
    console.log('AFTER', JSON.stringify(newAdb.tables, undefined, 2))
  }

  //Diff
  let ops = await computeDiff(existingAdb, newAdb, {
    updateComments: finalOptions.updateComments,
  })

  if (finalOptions.operationFilter) {
    ops = finalOptions.operationFilter.filter(ops);
  }

  if (finalOptions.debug) {
    // eslint-disable-next-line no-console
    console.log('OPERATIONS', ops)
  }

  //Write back to DB
  await write(
    ops,
    config,
    finalOptions.dbSchemaName,
    finalOptions.dbTablePrefix,
    finalOptions.dbColumnPrefix,
    finalOptions.plugins,
  )

  return { results: ops, previousDB: existingAdb, newDB: newAdb }
}
