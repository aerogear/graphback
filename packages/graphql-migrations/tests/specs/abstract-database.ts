/*eslint-disable max-lines*/
import { buildSchema } from 'graphql';
import { generateAbstractDatabase } from '../../src'

test('skip root types', async () => {
  const schema = buildSchema(`
      type Query {
        hello: String
      }

      type Mutation {
        do: Boolean
      }

      type Subscription {
        notif: String
      }
    `)
  const adb = await generateAbstractDatabase(schema)
  expect(adb.tables.length).toEqual(0);
})

test('simple type', async () => {
  const schema = buildSchema(`
      """
      A user.
      """
      type User {
        id: ID!
        """
        Display name.
        """
        name: String!
      }
    `)
  const adb = await generateAbstractDatabase(schema)
  expect(adb.tables.length).toEqual(1);
  const [User] = adb.tables
  expect(User.name).toEqual('user');
  expect(User.comment).toEqual('A user.');
  expect(User.columns.length).toEqual(2);
  const [colId, colName] = User.columns
  expect(colId.name).toEqual('id');
  expect(colId.type).toEqual('increments');
  expect(colName.name).toEqual('name');
  expect(colName.type).toEqual('string');
  expect(colName.comment).toEqual('Display name.');
})

test('skip table', async () => {
  const schema = buildSchema(`
      """
      @db.skip
      """
      type User {
        id: ID!
        name: String!
      }
    `)
  const adb = await generateAbstractDatabase(schema)
  expect(adb.tables.length).toEqual(0);
})

test('skip field', async () => {
  const schema = buildSchema(`
      type User {
        id: ID!
        """
        @db.skip
        """
        name: String!
      }
    `)
  const adb = await generateAbstractDatabase(schema)
  expect(adb.tables.length).toEqual(1);
  const [User] = adb.tables
  expect(User.name).toEqual('user');
  expect(User.columns.length).toEqual(1);
  const [colId] = User.columns
  expect(colId.name).toEqual('id');
})

test('not null', async () => {
  const schema = buildSchema(`
      type User {
        id: ID!
        name: String!
        nickname: String
      }
    `)
  const adb = await generateAbstractDatabase(schema)
  expect(adb.tables.length).toEqual(1);
  const [User] = adb.tables
  expect(User.columns.length).toEqual(3);
  const [_, colName, colNickname] = User.columns
  expect(colName.nullable).toBeFalsy();
  expect(colNickname.nullable).toBeTruthy();
})

test('default value', async () => {
  const schema = buildSchema(`
      type User {
        id: ID!
        """
        @db.default: true
        """
        someOption: Boolean
        """
        @db.default: false
        """
        thatOption: Boolean
        """
        @db.default: ''
        """
        thisOption: Boolean
      }
    `)
  const adb = await generateAbstractDatabase(schema)
  const [User] = adb.tables
  const [_, colSomeOption, colThatOption, colThisOption] = User.columns
  expect(colSomeOption.defaultValue).toBeTruthy();
  expect(colThatOption.defaultValue).toBeFalsy();
  expect(colThisOption.defaultValue).toEqual('');
})

test('default primary index', async () => {
  const schema = buildSchema(`
      type User {
        """
        This will get a primary index
        """
        id: ID!
        email: String!
      }
    `)
  const adb = await generateAbstractDatabase(schema)
  expect(adb.tables.length).toEqual(1);
  const [User] = adb.tables
  expect(User.primaries.length).toEqual(1);
  const [id] = User.primaries
  expect(id.columns).toEqual(['id'])
})

test('simple index', async () => {
  const schema = buildSchema(`
      type User {
        id: ID!
        """
        @db.index
        """
        email: String!
      }
    `)
  const adb = await generateAbstractDatabase(schema)
  expect(adb.tables.length).toEqual(1);
  const [User] = adb.tables
  expect(User.indexes.length).toEqual(1);
  const [email] = User.indexes
  expect(email.columns).toEqual(['email'])
})

test('multiple indexes', async () => {
  const schema = buildSchema(`
      type User {
        """
        @db.index
        """
        id: ID!
        """
        @db.index
        """
        email: String!
      }
    `)
  const adb = await generateAbstractDatabase(schema)
  expect(adb.tables.length).toEqual(1);
  const [User] = adb.tables
  expect(User.indexes.length).toEqual(2);
  const [id, email] = User.indexes
  expect(id.columns).toEqual(['id'])
  expect(email.columns).toEqual(['email'])
})

test('named index', async () => {
  const schema = buildSchema(`
      type User {
        id: ID!
        """
        @db.index: 'myIndex'
        """
        email: String!
        """
        @db.index: 'myIndex'
        """
        name: String!
      }
    `)
  const adb = await generateAbstractDatabase(schema)
  expect(adb.tables.length).toEqual(1);
  const [User] = adb.tables
  expect(User.indexes.length).toEqual(1);
  const [myIndex] = User.indexes
  expect(myIndex.name).toEqual('myIndex');;
  expect(myIndex.columns).toEqual(['email', 'name'])
})

test('object index', async () => {
  const schema = buildSchema(`
      type User {
        id: ID!
        """
        @db.index: { name: 'myIndex', type: 'string' }
        """
        email: String!
        """
        @db.index: 'myIndex'
        """
        name: String!
      }
    `)
  const adb = await generateAbstractDatabase(schema)
  expect(adb.tables.length).toEqual(1);
  const [User] = adb.tables
  expect(User.indexes.length).toEqual(1);
  const [myIndex] = User.indexes
  expect(myIndex.name).toEqual('myIndex');
  expect(myIndex.type).toEqual('string');
  expect(myIndex.columns).toEqual(['email', 'name'])
})

test('unique index', async () => {
  const schema = buildSchema(`
      type User {
        id: ID!
        """
        @db.unique
        """
        email: String!
      }
    `)
  const adb = await generateAbstractDatabase(schema)
  expect(adb.tables.length).toEqual(1);
  const [User] = adb.tables
  expect(User.uniques.length).toEqual(1);
  const [email] = User.uniques
  expect(email.columns).toEqual(['email'])
})

test('custom name', async () => {
  const schema = buildSchema(`
      """
      @db.name: 'people'
      """
      type User {
        id: ID!
      }
    `)
  const adb = await generateAbstractDatabase(schema)
  expect(adb.tables.length).toEqual(1)
  const [User] = adb.tables
  expect(User.annotations.name).toEqual('people')
  expect(User.name).toEqual('people')
})

test('custom type', async () => {
  const schema = buildSchema(`
      type User {
        id: ID!
        """
        @db.type: 'string'
        @db.length: 36
        """
        name: String
      }
    `)
  const adb = await generateAbstractDatabase(schema)
  expect(adb.tables.length).toEqual(1)
  const [User] = adb.tables
  const [_, colName] = User.columns
  expect(colName.name).toEqual('name')
  expect(colName.annotations.type).toEqual('string')
  expect(colName.annotations.length).toEqual(36)
  expect(colName.type).toEqual('string')
  expect(colName.args).toEqual([36])
})

test('foreign key', async () => {
  const schema = buildSchema(`
      type User {
        id: ID!
        messages: [Message!]!
      }

      type Message {
        id: ID!
        user: User
      }
    `)
  const adb = await generateAbstractDatabase(schema)
  expect(adb.tables.length).toEqual(2)
  const [User, Message] = adb.tables
  expect(User.name).toEqual('user')
  expect(Message.name).toEqual('message')
  expect(User.columns.length).toEqual(1)
  expect(Message.columns.length).toEqual(2)
  const [colId, colUserForeign] = Message.columns
  expect(colId.name).toEqual('id')
  expect(colUserForeign.name).toEqual('userId')
  expect(colUserForeign.type).toEqual('integer')
  expect(colUserForeign.foreign && colUserForeign.foreign.tableName === 'user').toBeTruthy()
  expect(colUserForeign.foreign && colUserForeign.foreign.columnName === 'id').toBeTruthy()
})

test('many to many', async () => {
  const schema = buildSchema(`
      type User {
        id: ID!
        """
        @db.manyToMany: 'users'
        """
        messages: [Message!]!
      }

      type Message {
        id: ID!
        """
        @db.manyToMany: 'messages'
        """
        users: [User]
      }
    `)
  const adb = await generateAbstractDatabase(schema)
  expect(adb.tables.length).toEqual(3)
  const Join = adb.tables[2]
  expect(Join.name).toEqual('message_users_join_user_messages');
  const [colMessageUsers, colUserMessages] = Join.columns
  expect(colMessageUsers.name).toEqual('usersId')
  expect(colMessageUsers.type).toEqual('integer')
  expect(colMessageUsers.foreign && colMessageUsers.foreign.tableName === 'message').toBeTruthy()
  expect(colMessageUsers.foreign && colMessageUsers.foreign.columnName === 'id').toBeTruthy()
  expect(colUserMessages.name).toEqual('messagesId')
  expect(colUserMessages.type).toEqual('integer')
  expect(colUserMessages.foreign && colUserMessages.foreign.tableName === 'user').toBeTruthy()
  expect(colUserMessages.foreign && colUserMessages.foreign.columnName === 'id').toBeTruthy()
})

test('many to many on self', async () => {
  const schema = buildSchema(`
      type User {
        id: ID!
        contacts: [User]
      }
    `)
  const adb = await generateAbstractDatabase(schema)
  expect(adb.tables.length).toEqual(2)
  const [User, UserContacts] = adb.tables
  expect(UserContacts.name).toEqual('user_contacts_join_user_contacts')
  expect(User.name).toEqual('user')
  expect(User.columns.length).toEqual(1)
  const [col1, col2] = UserContacts.columns
  expect(col1.name).toEqual('idId')
  expect(col1.type).toEqual('integer')
  expect(col1.foreign && col1.foreign.tableName === 'user').toBeTruthy()
  expect(col1.foreign && col1.foreign.columnName === 'id').toBeTruthy()
  expect(col2.name).toEqual('idId_other')
  expect(col2.type).toEqual('integer')
  expect(col2.foreign && col2.foreign.tableName === 'user').toBeTruthy()
  expect(col2.foreign && col2.foreign.columnName === 'id').toBeTruthy()
})

test('simple list', async () => {
  const schema = buildSchema(`
      type User {
        id: ID!
        """
        @db.type: 'json'
        """
        names: [String]
      }
    `)
  const adb = await generateAbstractDatabase(schema)
  expect(adb.tables.length).toEqual(1);
  const [User] = adb.tables
  expect(User.name).toEqual('user');
  expect(User.columns.length).toEqual(2);
  const [colId, colNames] = User.columns
  expect(colId.name).toEqual('id');
  expect(colId.type).toEqual('increments');
  expect(colNames.name).toEqual('names');
  expect(colNames.type).toEqual('json');
})

test('custom scalar map', async () => {
  const schema = buildSchema(`
      type User {
        id: ID
        name: String
        nickname: String
      }
    `)
  const adb = await generateAbstractDatabase(schema, {
    scalarMap: (field: any) => {
      if (field.name === 'name') {
        return {
          type: 'text',
          args: [],
        }
      }

      return undefined
    },
  })
  expect(adb.tables.length).toEqual(1)
  const [User] = adb.tables
  expect(User.columns.length).toEqual(3)
  const [colId, colName, colNickname] = User.columns
  expect(colId.type).toEqual('increments')
  expect(colName.type).toEqual('text')
  expect(colNickname.type).toEqual('string')
})

test('map lists to json', async () => {
  const schema = buildSchema(`
      type User {
        id: ID!
        names: [String]
      }
    `)
  const adb = await generateAbstractDatabase(schema, {
    mapListToJson: true,
  })
  expect(adb.tables.length).toEqual(1)
  const [User] = adb.tables
  expect(User.columns.length).toEqual(2)
  const [_, colNames] = User.columns
  expect(colNames.type).toEqual('json')
})

test('default name transforms', async () => {
  const schema = buildSchema(`
      type UserTeam {
        id: ID!
        TeamName: String!
        yearlyBilling: Boolean!
      }
    `)
  const adb = await generateAbstractDatabase(schema)
  expect(adb.tables.length).toEqual(1)
  const [UserTeam] = adb.tables
  expect(UserTeam.name).toEqual('userteam')
  expect(UserTeam.columns.length).toEqual(3)
  const [colId, colName, colYearlyBilling] = UserTeam.columns
  expect(colId.name).toEqual('id')
  expect(colName.name).toEqual('TeamName')
  expect(colYearlyBilling.name).toEqual('yearlyBilling')
})

test('sandbox', async () => {
  const schema = buildSchema(`
      scalar Date

      """
      A user.
      """
      type User {
        id: ID!
        """
        Display name
        @db.length: 200
        """
        name: String!
        email: String!
        score: Int
        """
        @db.type: 'json'
        """
        scores: [Int]
        messages: [Message]
        """
        @db.manyToMany: 'users'
        """
        sharedMessages: [Message]
        contacts: [User]
      }

      type Message {
        id: ID!
        user: User!
        """
        @db.manyToMany: 'sharedMessages'
        """
        users: [User]
        """
        @db.type: 'datetime'
        """
        created: Date!
        title: String!
        """
        @db.type: 'text'
        """
        content: String!
      }

      type Query {
        users: [User]
      }
    `)
  const adb = await generateAbstractDatabase(schema)
  expect(adb.tables.length).toEqual(4);
  expect(adb.tables).toMatchSnapshot();
})
