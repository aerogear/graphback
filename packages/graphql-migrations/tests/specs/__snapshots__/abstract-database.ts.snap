// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sandbox 1`] = `
Array [
  Object {
    "annotations": Object {},
    "columnMap": Map {
      "id" => Object {
        "annotations": Object {},
        "args": Array [],
        "autoIncrementable": true,
        "comment": undefined,
        "defaultValue": undefined,
        "foreign": undefined,
        "isPrimaryKey": true,
        "name": "id",
        "nullable": false,
        "type": "increments",
      },
      "name" => Object {
        "annotations": Object {
          "length": 200,
        },
        "args": Array [
          200,
        ],
        "autoIncrementable": false,
        "comment": "Display name",
        "defaultValue": undefined,
        "foreign": undefined,
        "isPrimaryKey": false,
        "name": "name",
        "nullable": false,
        "type": "string",
      },
      "email" => Object {
        "annotations": Object {},
        "args": Array [
          255,
        ],
        "autoIncrementable": false,
        "comment": undefined,
        "defaultValue": undefined,
        "foreign": undefined,
        "isPrimaryKey": false,
        "name": "email",
        "nullable": false,
        "type": "string",
      },
      "score" => Object {
        "annotations": Object {},
        "args": Array [],
        "autoIncrementable": false,
        "comment": undefined,
        "defaultValue": undefined,
        "foreign": undefined,
        "isPrimaryKey": false,
        "name": "score",
        "nullable": true,
        "type": "integer",
      },
      "scores" => Object {
        "annotations": Object {
          "type": "json",
        },
        "args": Array [],
        "autoIncrementable": false,
        "comment": undefined,
        "defaultValue": undefined,
        "foreign": undefined,
        "isPrimaryKey": false,
        "name": "scores",
        "nullable": true,
        "type": "json",
      },
    },
    "columns": Array [
      Object {
        "annotations": Object {},
        "args": Array [],
        "autoIncrementable": true,
        "comment": undefined,
        "defaultValue": undefined,
        "foreign": undefined,
        "isPrimaryKey": true,
        "name": "id",
        "nullable": false,
        "type": "increments",
      },
      Object {
        "annotations": Object {
          "length": 200,
        },
        "args": Array [
          200,
        ],
        "autoIncrementable": false,
        "comment": "Display name",
        "defaultValue": undefined,
        "foreign": undefined,
        "isPrimaryKey": false,
        "name": "name",
        "nullable": false,
        "type": "string",
      },
      Object {
        "annotations": Object {},
        "args": Array [
          255,
        ],
        "autoIncrementable": false,
        "comment": undefined,
        "defaultValue": undefined,
        "foreign": undefined,
        "isPrimaryKey": false,
        "name": "email",
        "nullable": false,
        "type": "string",
      },
      Object {
        "annotations": Object {},
        "args": Array [],
        "autoIncrementable": false,
        "comment": undefined,
        "defaultValue": undefined,
        "foreign": undefined,
        "isPrimaryKey": false,
        "name": "score",
        "nullable": true,
        "type": "integer",
      },
      Object {
        "annotations": Object {
          "type": "json",
        },
        "args": Array [],
        "autoIncrementable": false,
        "comment": undefined,
        "defaultValue": undefined,
        "foreign": undefined,
        "isPrimaryKey": false,
        "name": "scores",
        "nullable": true,
        "type": "json",
      },
    ],
    "comment": "A user.",
    "indexes": Array [],
    "name": "user",
    "uniques": Array [],
  },
  Object {
    "annotations": Object {},
    "columnMap": Map {
      "id" => Object {
        "annotations": Object {},
        "args": Array [],
        "autoIncrementable": true,
        "comment": undefined,
        "defaultValue": undefined,
        "foreign": undefined,
        "isPrimaryKey": true,
        "name": "id",
        "nullable": false,
        "type": "increments",
      },
      "user" => Object {
        "annotations": Object {},
        "args": Array [],
        "autoIncrementable": false,
        "comment": undefined,
        "defaultValue": undefined,
        "foreign": Object {
          "columnName": "id",
          "field": "id",
          "tableName": "user",
          "type": "User",
        },
        "isPrimaryKey": false,
        "name": "userId",
        "nullable": false,
        "type": "integer",
      },
      "created" => Object {
        "annotations": Object {
          "type": "datetime",
        },
        "args": Array [
          undefined,
        ],
        "autoIncrementable": false,
        "comment": undefined,
        "defaultValue": undefined,
        "foreign": undefined,
        "isPrimaryKey": false,
        "name": "created",
        "nullable": false,
        "type": "datetime",
      },
      "title" => Object {
        "annotations": Object {},
        "args": Array [
          255,
        ],
        "autoIncrementable": false,
        "comment": undefined,
        "defaultValue": undefined,
        "foreign": undefined,
        "isPrimaryKey": false,
        "name": "title",
        "nullable": false,
        "type": "string",
      },
      "content" => Object {
        "annotations": Object {
          "type": "text",
        },
        "args": Array [],
        "autoIncrementable": false,
        "comment": undefined,
        "defaultValue": undefined,
        "foreign": undefined,
        "isPrimaryKey": false,
        "name": "content",
        "nullable": false,
        "type": "text",
      },
    },
    "columns": Array [
      Object {
        "annotations": Object {},
        "args": Array [],
        "autoIncrementable": true,
        "comment": undefined,
        "defaultValue": undefined,
        "foreign": undefined,
        "isPrimaryKey": true,
        "name": "id",
        "nullable": false,
        "type": "increments",
      },
      Object {
        "annotations": Object {},
        "args": Array [],
        "autoIncrementable": false,
        "comment": undefined,
        "defaultValue": undefined,
        "foreign": Object {
          "columnName": "id",
          "field": "id",
          "tableName": "user",
          "type": "User",
        },
        "isPrimaryKey": false,
        "name": "userId",
        "nullable": false,
        "type": "integer",
      },
      Object {
        "annotations": Object {
          "type": "datetime",
        },
        "args": Array [
          undefined,
        ],
        "autoIncrementable": false,
        "comment": undefined,
        "defaultValue": undefined,
        "foreign": undefined,
        "isPrimaryKey": false,
        "name": "created",
        "nullable": false,
        "type": "datetime",
      },
      Object {
        "annotations": Object {},
        "args": Array [
          255,
        ],
        "autoIncrementable": false,
        "comment": undefined,
        "defaultValue": undefined,
        "foreign": undefined,
        "isPrimaryKey": false,
        "name": "title",
        "nullable": false,
        "type": "string",
      },
      Object {
        "annotations": Object {
          "type": "text",
        },
        "args": Array [],
        "autoIncrementable": false,
        "comment": undefined,
        "defaultValue": undefined,
        "foreign": undefined,
        "isPrimaryKey": false,
        "name": "content",
        "nullable": false,
        "type": "text",
      },
    ],
    "comment": undefined,
    "indexes": Array [],
    "name": "message",
    "uniques": Array [],
  },
  Object {
    "annotations": Object {},
    "columnMap": Map {
      "usersId" => Object {
        "annotations": Object {},
        "args": Array [],
        "autoIncrementable": false,
        "comment": undefined,
        "defaultValue": undefined,
        "foreign": Object {
          "columnName": "id",
          "field": "id",
          "tableName": "message",
          "type": "Message",
        },
        "isPrimaryKey": false,
        "name": "usersId",
        "nullable": true,
        "type": "integer",
      },
      "sharedMessagesId" => Object {
        "annotations": Object {},
        "args": Array [],
        "autoIncrementable": false,
        "comment": undefined,
        "defaultValue": undefined,
        "foreign": Object {
          "columnName": "id",
          "field": "id",
          "tableName": "user",
          "type": "User",
        },
        "isPrimaryKey": false,
        "name": "sharedMessagesId",
        "nullable": true,
        "type": "integer",
      },
    },
    "columns": Array [
      Object {
        "annotations": Object {},
        "args": Array [],
        "autoIncrementable": false,
        "comment": undefined,
        "defaultValue": undefined,
        "foreign": Object {
          "columnName": "id",
          "field": "id",
          "tableName": "message",
          "type": "Message",
        },
        "isPrimaryKey": false,
        "name": "usersId",
        "nullable": true,
        "type": "integer",
      },
      Object {
        "annotations": Object {},
        "args": Array [],
        "autoIncrementable": false,
        "comment": undefined,
        "defaultValue": undefined,
        "foreign": Object {
          "columnName": "id",
          "field": "id",
          "tableName": "user",
          "type": "User",
        },
        "isPrimaryKey": false,
        "name": "sharedMessagesId",
        "nullable": true,
        "type": "integer",
      },
    ],
    "comment": "[Auto] Join table between User.sharedMessages and Message.users",
    "indexes": Array [
      Object {
        "columns": Array [
          "usersId",
        ],
        "name": "message_users_join_user_sharedmessages_usersId_index",
        "type": undefined,
      },
      Object {
        "columns": Array [
          "sharedMessagesId",
        ],
        "name": "message_users_join_user_sharedmessages_sharedMessagesId_index",
        "type": undefined,
      },
    ],
    "name": "message_users_join_user_sharedmessages",
    "uniques": Array [],
  },
  Object {
    "annotations": Object {},
    "columnMap": Map {
      "idId" => Object {
        "annotations": Object {},
        "args": Array [],
        "autoIncrementable": false,
        "comment": undefined,
        "defaultValue": undefined,
        "foreign": Object {
          "columnName": "id",
          "field": "id",
          "tableName": "user",
          "type": "User",
        },
        "isPrimaryKey": false,
        "name": "idId",
        "nullable": false,
        "type": "integer",
      },
      "idId_other" => Object {
        "annotations": Object {},
        "args": Array [],
        "autoIncrementable": false,
        "comment": undefined,
        "defaultValue": undefined,
        "foreign": Object {
          "columnName": "id",
          "field": "id",
          "tableName": "user",
          "type": "User",
        },
        "isPrimaryKey": false,
        "name": "idId_other",
        "nullable": false,
        "type": "integer",
      },
    },
    "columns": Array [
      Object {
        "annotations": Object {},
        "args": Array [],
        "autoIncrementable": false,
        "comment": undefined,
        "defaultValue": undefined,
        "foreign": Object {
          "columnName": "id",
          "field": "id",
          "tableName": "user",
          "type": "User",
        },
        "isPrimaryKey": false,
        "name": "idId",
        "nullable": false,
        "type": "integer",
      },
      Object {
        "annotations": Object {},
        "args": Array [],
        "autoIncrementable": false,
        "comment": undefined,
        "defaultValue": undefined,
        "foreign": Object {
          "columnName": "id",
          "field": "id",
          "tableName": "user",
          "type": "User",
        },
        "isPrimaryKey": false,
        "name": "idId_other",
        "nullable": false,
        "type": "integer",
      },
    ],
    "comment": "[Auto] Join table between User.contacts and User.contacts",
    "indexes": Array [
      Object {
        "columns": Array [
          "idId",
          "idId_other",
        ],
        "name": "user_contacts_join_user_contacts_idId_idId_other_index",
        "type": undefined,
      },
    ],
    "name": "user_contacts_join_user_contacts",
    "uniques": Array [],
  },
]
`;
